---
title: Running Destin on 10X Genomics scATAC-seq data of 5k PBMC cells
author: Eugene Urrutia, Yuchao Jiang
output: 
    html_document:
        theme: cosmo 
        toc: true
        toc_float: true
        highlight: tango
        number_sections: false
fig_width: 5
fig_height: 5 
---

# Introduction

In this example we use the following data set: 

5k Peripheral blood mononuclear cells (PBMCs) from a healthy donor

Single Cell ATAC Dataset by Cell Ranger ATAC 1.0.1

Peripheral blood mononuclear cells (PBMCs) from a healthy donor.

- ~6700 transposed nuclei were loaded.
- 5335 nuclei were recovered.
- Sequenced on Illumina NovaSeq with approximately 42k read pairs per cell.
- 50bp read1, 8bp i7 (sample index), 16bp i5 (10x Barcode), 49bp read2.
- Published on December 17, 2018

This dataset is licensed under the Creative Commons Attribution license.

https://support.10xgenomics.com/single-cell-atac/datasets/1.0.1/atac_v1_pbmc_5k

10x uses Cell Ranger ATAC 1.0.1 for bioinformatic processing and produces a peak by cell matrix. 

Approximate time to run Destin on this dataset (8 core macbook pro): ~ 15 min

```{r}
startTime = Sys.time()
```

# Download Data

Download the file corresponding to Peak by cell matrix (filtered)	97.86 MB	03ea4420a7bdaadd8c1cab8d4a0cf2e9

We download the peak by cell matrix via the terminal using curl, although other options are available on thier webpage. 

```{bash eval=F}
cd localDir10xData
curl -O http://cf.10xgenomics.com/samples/cell-atac/1.0.1/atac_v1_pbmc_5k/atac_v1_pbmc_5k_filtered_peak_bc_matrix.tar.gz
```

Unpack the tar.gz download via the terminal (method may depend on operating system)

```{bash eval=F}
tar -xzf atac_v1_pbmc_5k_filtered_peak_bc_matrix.tar.gz
```

The resulting folder contains 3 files:

- `matrix.mtx`:  peak by cell chromatin accessiblilty matrix in MatrixMarket format.  We can use R package "Matrix" to read straight to a sparse matrix
- `peaks.bed`: bed file corresponding to rows of matrix.mtx
- `barcodes.tsv`:  barcodes corresponding to columns of matrix.mtx

set data10xDir to appropriate local folder
```{r}
# data10xDir = "~/Dropbox/GitHub/PBMC_10X/filtered_peak_bc_matrix"
data10xDir = "~/Dropbox/Documents/statGen/scATACseq/10xgenomics/atac_v1_pbmc_5k/filtered_peak_bc_matrix" 
```

install destin R package
```{r  message=FALSE, eval=FALSE}
yourPathToDestinRepo = "~/Documents/gitRepos/destin"
install.packages(file.path(yourPathToDestinRepo,"package"), repos = NULL, type = "source")
```

load Destin
```{r  message=FALSE, warning=FALSE}
library(destin, quietly = T) 
```

# Create ranged summarized experiment from 10x peak by cell matrix 
```{r}
rse = createRSEfrom10xMatrix(data10xDir)
```


# test memory limits; testing purposes only
```{r, eval = F}
countMat = assay(rse)
countMatBig = cbind(countMat, countMat, countMat, countMat, countMat) 
countMatBigger = cbind(countMatBig, countMatBig, countMatBig)
pca = irlba(t(countMatBigger), nv = 2)
```


here is a description of the ranged summarized experiment  

column data is a description of the cells
```{r}
colData(rse)
```

rowRanges is a description of the peaks
```{r}
rowRanges(rse)
```

assay is the binary chromatin accessibility matrix in sparse matrix format

- rows are peaks 
- columns are cells

```{r}
assay(rse)[1:10,1:10]
```

# Peak annotation
Destin calculates the distance of each peak to the transcript start site (TSS) and correspondingly annotates the peak to be distal or proximal regulatory element based on a 3kb window centered at the TSS. Destin also calculates for each peak its openness frequency (i.e., the frequency of this peak being open among existing cell lines and tissue types) based on the reference frequency map, pre-constructed from the ENCODE Project.

```{r}
  model = "hg19" # choose from hg19, hg38, mm10
  rse = annotateRSE(rse, model)
```

# Quality control

Destin begins with quality control, retaining chromatin regions accessible in at least 5 cells, and retaining cells with log total number of chromatin accessible regions within 3 standard deviations from median (robustly calculated using median absolute deviations). These default parameters can be modified if desired.

```{r}
  rse = doQC(rse, regionSumCutoff = 5, cellSumCutoffSDs = 3)
```

# Estimate number of clusters 

Destin uses an unsupervised model-based likelihood (logLikeElbow) to estimate number of clusters. Like WCSSE, the likelihood is non-decreasing with increasing number of clusters. We fit the model log-likelihood as a linear spline function of number of clusters with a single knot. The knot placement which results in the best fit (lowest sum of squared errors) is selected as number of clusters (n_k).

```{r, message=FALSE}
clusterEst = estimateNClusters(rse, nClustersRange = 2:20)
nClusters = clusterEst$nClustersList$logLikeElbow
plotNClusters(clusterEst)
```

# Perform Destin clustering 
```{r}
#set nCores to an appropriate value based on your system
nCores = 7
clusterResults = destinGrid (rse, nClusters = nClusters, nCores = nCores)
clusterResults$cluster
```

# Plot results
```{r warning=FALSE}
plotCluster(clusterResults, type = "t-SNE")
```

Time to run
```{r}
TotalTime = Sys.time() - startTime
TotalTime
```
